src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,482 @@
+		return elliptic.P521()
+	default:
+		return nil
+	}
+}
+
+// prepareDelegation returns the message that the delegator is going to sign.
+// The inputs are the credential ('cred'), the DER-encoded end-entity
+// certificate ('dCert'), the signature scheme of the delegator
+// ('algo').
+func prepareDelegation(cred *Credential, dCert []byte, algo SignatureScheme, peer string) ([]byte, error) {
+	values := make([]byte, 64, 128)
+	for i := range values {
+		values[i] = 0x20
+	}
+
+	if peer == "server" {
+		values = append(values, []byte("TLS, server delegated credentials\x00")...)
+	} else if peer == "client" {
+		values = append(values, []byte("TLS, client delegated credentials\x00")...)
+	}
+
+	values = append(values, dCert...)
+
+	serCred, err := cred.marshal()
+	if err != nil {
+		return nil, err
+	}
+	values = append(values, serCred...)
+
+	var serAlgo [2]byte
+	binary.BigEndian.PutUint16(serAlgo[:], uint16(algo))
+	values = append(values, serAlgo[:]...)
+
+	return values, nil
+}
+
+// NewDelegatedCredential creates a new delegated credential using 'cert' for
+// delegation. It generates a public/private key pair for the provided signature
+// algorithm ('scheme'), validity interval (defined by 'cert.Leaf.notBefore' and
+// 'validTime'), and TLS version ('vers'), and signs it using 'cert.PrivateKey'.
+// TODO: add the other signature schemes
+func NewDelegatedCredential(cert *Certificate, pubAlgo SignatureScheme, validTime time.Duration, peer string) (*DelegatedCredential, crypto.PrivateKey, error) {
+	// The granularity of DC validity is seconds.
+	validTime = validTime.Round(time.Second)
+
+	// Parse the leaf certificate if needed.
+	var err error
+	if cert.Leaf == nil {
+		if len(cert.Certificate[0]) == 0 {
+			return nil, nil, errors.New("tls: missing leaf certificate for delegated credential")
+		}
+		cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+		if err != nil {
+			return nil, nil, err
+		}
+
+	}
+
+	// Check that the leaf certificate can be used for delegation.
+	if !isValidForDelegation(cert.Leaf) {
+		return nil, nil, errNoDelegationUsage
+	}
+
+	// Extract the algorithm used to sign the DelegatedCredential from the end-entity (leaf) certificate
+	var sigAlgo SignatureScheme
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		pk := sk.Public().(*ecdsa.PublicKey)
+		curveName := pk.Curve.Params().Name
+		certAlg := cert.Leaf.SignatureAlgorithm
+		if certAlg == x509.ECDSAWithSHA256 && curveName == "P-256" {
+			sigAlgo = ECDSAWithP256AndSHA256
+		} else if certAlg == x509.ECDSAWithSHA384 && curveName == "P-384" {
+			sigAlgo = ECDSAWithP384AndSHA384
+		} else if certAlg == x509.ECDSAWithSHA512 && curveName == "P-521" {
+			sigAlgo = ECDSAWithP521AndSHA512
+		} else {
+			return nil, nil, fmt.Errorf("using curve %s for %s is not supported", curveName, cert.Leaf.SignatureAlgorithm)
+		}
+	default:
+		return nil, nil, fmt.Errorf("unsupported algorithm for delegated credential: %T", sk)
```

Comment:
prevent leakage of `sk`, if someone replaces `%T` by `%v`.
```suggestion
		return nil, nil, fmt.Errorf("unsupported algorithm for delegated credential")
```